R/subset_datarqst.R

```diff
@@ -37,66 +45,141 @@
 
     output_individual <- standard_data$Individual_data %>%
-      dplyr::mutate(Pop_sp = paste(PopID, Species, sep = "_")) %>%
-      dplyr::filter(Pop_sp %in% filter)
+      dplyr::mutate(Pop_sp = paste(.data$PopID, .data$Species, sep = "_")) %>%
+      dplyr::filter(.data$Pop_sp %in% filter) %>%
+      dplyr::select(-.data$Pop_sp)
 
     output_location <- standard_data$Location_data %>%
-      dplyr::filter(PopID %in% unique_pops)
+      dplyr::filter(.data$PopID %in% unique_pops)
 
   } else {
 
+    if(include_conflicting){
+      #Species <- c(Species, 'CCCCCC') # Use after pipelines updated to standard format 1.1
+      Species <- c(Species, 'CCCCCC', 'CONFLICTED')
+    }
+
     output_brood <- standard_data$Brood_data %>%
-      dplyr::filter(PopID %in% {{PopID}} & Species %in% {{Species}})
+      dplyr::filter(.data$PopID %in% {{PopID}} & .data$Species %in% {{Species}})
 
     output_capture <- standard_data$Capture_data %>%
-      dplyr::filter(CapturePopID %in% {{PopID}} & Species %in% {{Species}})
+      dplyr::filter(.data$CapturePopID %in% {{PopID}} & .data$Species %in% {{Species}})
 
     output_individual <- standard_data$Individual_data %>%
-      dplyr::filter(PopID %in% {{PopID}} & Species %in% {{Species}})
+      dplyr::filter(.data$PopID %in% {{PopID}} & .data$Species %in% {{Species}})
 
     if(!all(unique(Species_codes$Code) %in% Species) & all(unique(pop_names$code) %in% PopID)){
 
       PopID <- pop_species_combos %>%
-        dplyr::filter(species %in% Species) %>%
-        dplyr::pull(pop) %>%
+        dplyr::filter(.data$species %in% Species) %>%
+        dplyr::pull(.data$pop) %>%
         unique()
 
     }
 
     output_location <- standard_data$Location_data %>%
-      dplyr::filter(PopID %in% {{PopID}})
+      dplyr::filter(.data$PopID %in% {{PopID}})
 
   }
 
-  message("Output subset as .csv files...")
+  #If keeping conflicted species: remove IndvIDs that do not appear in capture data from individual data
+  #(this ensures that only data on individuals identified at least once as one of the species of interest are retained)
+  if(include_conflicting){
+    output_individual <- output_individual %>%
+      dplyr::filter(!(.data$Species %in% c('CONFLICTED', 'CCCCCC') & !(.data$IndvID %in% output_capture$IndvID)))
+  }
 
-  save_path <- dirname(file)
+  #If removing conflicted species: remove IndvIDs that do not appear in individual data from capture data
+  #(this ensures that individuals identified as different species do not appear in capture data at all, even if
+  #they were identified as one of the species of interest at one or several captures.)
+  if(!include_conflicting){
+    output_capture <- output_capture %>%
+      dplyr::filter(.data$IndvID %in% output_individual$IndvID)
+  }
+  # NOTE: This comes with an assumption that - aside from individuals with conflicting species - all
+  #       individuals that appear in capture data also appear in individual data (they should)
```

Comment:
One thing that's still missing here is if an individual is conflicted but only one of the two conflicted species was included in the filters. In this case we will only return part of the capture records for that individual (i.e. only those where it was classified as the species on interest). This doesn't really make sense to me. If a user says they want to include conflicted individuals I would interpret that to mean they want all records for an individual where species is uncertain (regardless of what species it was classified as at the time).

Reply:
I think you could overcome this by first returning a vector of individual IDs from the `Individual data` table and then use individual ID to filter capture ID, rather than using PopID and Species. That way all records for a conflicted species will be returned.

Reply:
@LiamDBailey I have followed your suggestion, and the filtering now works as follows: 
1) Filter Individual data (if `include_conflicted = TRUE` then *all* individuals with `Species = 'CCCCCC'`or `CONFLICTED` will remain in the data)
2) Filter Capture data using the `IndvID`still present in Individual data
3) Identify all conflicted species individuals that were *never* observed as one of the species of interest, and remove them from the data. 

Step 3) is there to make sure that if someone requested data on blue tits and great tits, including conflicting species individuals, their output data will contain all individuals that have been identified at least once as a great tit or a blue tit, but it will not include individuals that have never been identified as either of this species (so an individual once recorded as a pied flycatcher, and once as a collared flycatcher, would not be included in the output data). 

Personally, I think this does make sense. 
Alternatively, we could just return all conflicting species individuals (irrespective of what they were identified as) in all cases. 
What do you think makes most sense?

Reply:
I think that makes sense. From the users perspective when they ask for (for example) great tit and conflicted species I would take that to mean "return all individuals that could _possibly_ be great tit". So an individual that has at one time been recorded as a great tit is relevant (even if it was also recorded as something else), but an individual only ever recorded as pied/collared flycatcher is not.

Reply:
Great. I have added code to make sure that will be handled as discussed, and further specified this in the function documentation.

Reply:
I was wondering if it will return also the species name which it was in conflict with? For example for conflicted great tit, return in other column (?) that it was also stated as blue tit in certain cases? Would that be of interest of users or would  they get this information in some other place?

Reply:
@zuzulaz which "potential" species an individual has been recorded as is not part of Individual data. It would be a bit cumbersome to have in a single or multiple columns, since - in theory - an individual could be observed as not just two but many different species (hopefully not!). 
However, if the user goes into Capture data and looks for the conflicted species individual there, then each capture will be associated with a species. 

For example: 

In Individual data, the species of individual X is denoted as "CCCCCC" .
In Capture data, individual X has two captures, one of which has Species = "PARMAJ" and one has Species = "CYACAE".

Reply:
Right! Thank you for reminding me that that information is in the capture data.